### PR TITLE
feat: abort controller

### PR DESCRIPTION
**Pull request by kwasniew**
## About the changes

Abort controller that aborts previous request when a new one is made.

Sample use cases:
* slow network + multiple consecutive updateContext calls
* slow network + very short interval

### CODE REVIEW

File: src/index.test.ts [hunk 1]
```diff
@@ -576,6 +576,25 @@ test('Should publish error when fetch fails', (done) => {
...
+        clientKey: '12',
+        appName: 'web',
+    };
+    const client = new UnleashClient(config);
+    await client.start();
+    client.updateContext({ userId: '123' }); // abort 1
+    client.updateContext({ userId: '456' }); // abort 2
+    await client.updateContext({ userId: '789' });
+
+    expect(abortSpy).toBeCalledTimes(2);
```
Comment (kwasniew) [inline src/index.test.ts:594]: I also wanted to test error emitter on abort but the mocking fetch library does not support AbortController

File: src/index.ts [hunk 2]
```diff
@@ -89,6 +90,18 @@ export const resolveFetch = () => {
     return undefined;
 };
 
+const resolveAbortController = () => {
```
Comment (kwasniew) [inline src/index.ts:93]: similar pattern to how we handle fetch

**Review by sighphyre — APPROVED** [review]
Yes! This is very nice

File: src/index.ts [hunk 3]
```diff
@@ -94,6 +95,18 @@ export const resolveFetch = () => {
...
+    try {
+        if (typeof window !== 'undefined' && 'AbortController' in window) {
+            return () => new window.AbortController();
+        } else if ('fetch' in globalThis) {
+            return () => new globalThis.AbortController();
+        }
+    } catch (e) {
+        console.error('Unleash failed to resolve "AbortController" factory', e);
```
Comment (sighphyre) [inline src/index.ts:106]: This should basically never happen, right? This is if someone is doing something unusual like swapping out fetch?

Comment (kwasniew) [reply src/index.ts:106]: Correct - someone would need to take deliberate effort to end up in this situation

**Comment by nammi-kobler** [comment]
I keep getting these:

image

the requests comes in pairs (within 2 seconds), and if I upgrade to v3.3.0 I get these every other request (so one is always `cached` and the other has this error in the console:

image

downgrading to 3.2.0 fixes these for me.

this is what Stack Overflow tells me about the `NS_BINDING_ABORTED` ones:
https://stackoverflow.com/questions/704561/ns-binding-aborted-shown-in-firefox-with-httpfox

**Comment by kwasniew** [comment]
@nammi-kobler those errors are expected when the requests are aborted e,g, when you're on a slow network and you have very short refreshInterval and the original request hasn't finished and you start the next one. From what I understand your refresh interval is 2s. Can you increase it and see if the problem persists? The abort operation should be an exception not a regular operation.

**Comment by nammi-kobler** [comment]
The refresh interval is set to15s, but for some reason they always come in pairs (every 15 sec), and then one of them always fails. any idea on why they would come in pairs?

**Comment by kwasniew** [comment]
@nammi-kobler Can you check if this PR https://github.com/Unleash/unleash-proxy-client-js/pull/207 fixes your issue? It's applied in the https://github.com/Unleash/unleash-proxy-client-js/releases/tag/v3.4.0 release

**Comment by nammi-kobler** [comment]
> @nammi-kobler Can you check if this PR #207 fixes your issue? It's applied in the https://github.com/Unleash/unleash-proxy-client-js/releases/tag/v3.4.0 release

Thanks for this!
Firefox:
image

Vivaldi:
image

Safari:
image

Looks better, but it seems like it still happens over time

**Comment by kwasniew** [comment]
@nammi-kobler Thanks for checking. So is it correct that the console.error is gone and now you're only left with the double request problem? Can you remind me if you use unleash-proxy-client-js directly or through react or nextjs wrapper? I couldn't reproduce this problem in this vanilla client so my guess is it's some wrapper adding those extra calls.

**Comment by nammi-kobler** [comment]
> @nammi-kobler Thanks for checking. So is it correct that the console.error is gone and now you're only left with the double request problem? Can you remind me if you use unleash-proxy-client-js directly or through react or nextjs wrapper? I couldn't reproduce this problem in this vanilla client so my guess is it's some wrapper adding those extra calls.

we are using `unleash-proxy-client`

image

NextJS warning is from a wrapper that comes with Mantine I think?

**Comment by kwasniew** [comment]
Are you running in Strict Mode by any chance? https://react.dev/reference/react/StrictMode
It runs code twice in development.

**Comment by nammi-kobler** [comment]
> Are you running in Strict Mode by any chance? https://react.dev/reference/react/StrictMode It runs code twice in development.

haha yes. that's maybe the answer then.

**Pull request merged**